src/qgraphicsmozview_p.cpp

```diff
@@ -295,10 +303,7 @@ void QGraphicsMozViewPrivate::OnLoadStarted(const char* aLocation)
 {
     Q_UNUSED(aLocation);
 
-    if (mIsPainted) {
-        mIsPainted = false;
-        mViewIface->firstPaint(-1, -1);
-    }
+    ResetPainted();
```

Comment:
The OnFirstPaint is called once per web site (per location and also counting reload) in WebBrowserCrhome and to me it feels like a correct and meaningful approach. However, when location change starts e.g. via QuickMozView::goBack we are not resetting isPainted property to false before EmbedLiteView::GoBack is called. Thus, there is a time slot starting from EmbedLiteView::GoBack to EmbedLiteViewListener::OnLoadStarted callback when we are falsely in painted state.

So, if we would remove lines 298-301 we would always stick in painted state even if location changes.

In addition, as there is unknown time span from EmbedLiteView::GoBack to the point when EmbedLiteViewListener::OnFirstPaint callback gets called, I would rather reset painted state before we actually invoke EmbedLiteView::GoBack (similarly for QuickMozView::goForward, QuickMozView::reload, and QuickMozView::load).

We cannot grab unless we have content painted.

All changes are not feasible to do from outside as goBack, goForward, reload, load are not virtual methods and imo those should not be virtual. I could re-implement mentioned methods in browser side by having small wrappers that would reset "painted state" and listen only firstPaint signal but that looks to me like a wrong approach as we have the "painted" property declared over here.